src/NServiceBus.Core/Recoverability/FirstLevelRetries/FailureInfoStorage.cs

```diff
@@ -0,0 +1,88 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+
+    // The data structure has fixed maximum size. When the data structure reaches its maximum size,
+    // the least recently used (LRU) message processing failure is removed from the storage.
+    class FailureInfoStorage
+    {
+        public FailureInfoStorage(int maxElements = 1000)
+        {
+            this.maxElements = maxElements;
+        }
+
+        public void RecordFailureInfoForMessage(string messageId, Exception exception)
+        {
+            lock (lockObject)
```

Comment:
> Synchronization Preference Order
If you decide you do need some sort of synchronization, then understand that not all of them have equal
performance or behavior characteristics. If performance is not critical, just use lock for simplicity and
clarity. Using anything other than a lock should require some intense measurement to justify the additional
complexity. In general, consider synchronization mechanisms in this order:
1. No synchronization at all
2. Simple Interlocked methods
3. lock/Monitor class
4. Asynchronous locks (see later in this chapter)
5. Everything else

> ...
> Try to avoid locking if at all possible, even if you need to significantly restructure your code. When
needed, use the simplest locking possible. For simple state changes, use Interlocked methods. For short or
rarely contended sections, use lock/Monitor and use a private field as the synchronization object. If you
have a highly contended critical section that lasts more than a few milliseconds, consider using the
asynchronous locking pattern like SemaphoreSlim.

About concurrent collections

> They are convenient, but you need to be careful—every single access to the collection involves
synchronization. Often, that is overkill and could harm your performance if there is a lot of contention. If
you have necessarily “chatty” read/write access patterns to the collection, this low-level synchronization
may make sense.

From http://www.writinghighperf.net/